src/publisher/mqtt_publisher.py

```diff
@@ -27,111 +31,141 @@ def __init__(self, configuration: Configuration) -> None:
         self.last_charge_state_by_vin: dict[str, str] = {}
         self.vin_by_charger_connected_topic: dict[str, str] = {}
         self.first_connection = True
+        self.client: None | aiomqtt.Client = None
+        self.__running: asyncio.Task[None] | None = None
+        self.__connected = asyncio.Event()
 
-        mqtt_client = gmqtt.Client(
-            client_id=str(self.publisher_id),
-            transport=self.transport_protocol.transport_mechanism,
-            will_message=gmqtt.Message(
-                topic=self.get_topic(mqtt_topics.INTERNAL_LWT, False),
-                payload="offline",
-                retain=True,
-            ),
-        )
-        mqtt_client.on_connect = self.__on_connect
-        mqtt_client.on_message = self.__on_message
-        self.client: Final[gmqtt.Client] = mqtt_client
-
-    @override
-    async def connect(self) -> None:
-        if self.configuration.mqtt_user is not None:
-            if self.configuration.mqtt_password is not None:
-                self.client.set_auth_credentials(
-                    username=self.configuration.mqtt_user,
-                    password=self.configuration.mqtt_password,
-                )
-            else:
-                self.client.set_auth_credentials(username=self.configuration.mqtt_user)
-
+    async def __run_loop(self) -> None:
+        if not self.host:
+            LOG.info("MQTT host is not configured")
+            return
+        ssl_context: ssl.SSLContext | None = None
         if self.transport_protocol.with_tls:
             ssl_context = ssl.create_default_context()
-            cert_uri = self.configuration.tls_server_cert_path
-            if cert_uri:
-                LOG.debug(f"Using custom CA file {cert_uri}")
-                ssl_context.load_verify_locations(cafile=cert_uri)
+            if self.configuration.tls_server_cert_path:
+                LOG.debug(
+                    f"Using custom CA file {self.configuration.tls_server_cert_path}"
+                )
+                ssl_context.load_verify_locations(
+                    cafile=self.configuration.tls_server_cert_path
+                )
                 if not self.configuration.tls_server_cert_check_hostname:
                     LOG.warning(
                         f"Skipping hostname check for TLS connection to {self.host}"
                     )
-                    ssl_context.check_hostname = False
-        else:
-            ssl_context = None
-        await self.client.connect(
-            host=self.host,
+
+        client = aiomqtt.Client(
+            hostname=self.host,
             port=self.port,
-            version=gmqtt.constants.MQTTv311,
-            ssl=ssl_context,
+            identifier=str(self.publisher_id) + "a",
+            transport=self.transport_protocol.transport_mechanism,
+            username=self.configuration.mqtt_user or None,
+            password=self.configuration.mqtt_password or None,
+            clean_session=True,
```

Comment:
clean_session, currently set to True to maintain current behaviour

In a future PR this could be changed to clean up subscribe logic on reconnect

clean_session – If True, the broker will remove all information about this client when it disconnects. If False, the client is a persistent client and subscription information and queued messages will be retained when the client disconnects.